File: 0804_unique_more_code_words/unique_morse.py
from typing import List


def uniqueMorseRepresentations(words: List[str]):
    """
    Time: O(s) where S is the sum of the length of words
    Space: O(s)
    """
    MORSE = [".-","-...","-.-.","-..",".","..-.","--.",
             "....","..",".---","-.-",".-..","--","-.",
             "---",".--.","--.-",".-.","...","-","..-",
             "...-",".--","-..-","-.--","--.."]
    seen = {"".join(MORSE[ord(c) - ord('a')] for c in word)
            for word in words}
    return len(seen)

File: 0804_unique_more_code_words/unique_morse.java
class Solution {
    // Time: O(s) where S is the sum of the length of words
    // Space: O(s)
    public int uniqueMorseRepresentations(String[] words) {
        String[] MORSE = new String[]{
            ".-","-...","-.-.","-..",".","..-.","--.",
            "....","..",".---","-.-",".-..","--","-.",
            "---",".--.","--.-",".-.","...","-","..-",
            "...-",".--","-..-","-.--","--.."};
        Set<String> seen = new HashSet();
        for (String word: words) {
            StringBuilder code = new StringBuilder();
            for (char c: word.toCharArray())
                code.append(MORSE[c - 'a']);
            seen.add(code.toString());
        }
        return seen.size();
    }
}